### PR TITLE
Template part selection component - fix keyboard controls.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
## Description

Enables keyboard controls to be used in the template part selection component.

The template part selector is not currently set up to properly handle keyboard controls.  When a user presses keys (like "up" or "down" for example), these events bubble out to other control systems in the editor closing the dropdown and selecting other blocks.  Even if these events did not bubble, there is no system in place to allow arrow keys to be used to navigate through the list of available items.  This PR:
* Stops key events from bubbling out of the selection component.
* Uses `composite` from reakit to enable keyboard controls (as also used in the `BlockListTypes`)

## How has this been tested?

Template part selection component should be keyboard navigable similar to how the block inserter is:
* Tab / shft+tab - should toggle focus between the search input and the list results
* Arrow keys:
   * In search input - should no longer close the selection component
   * in the list - should allow navigation within the list results.
* Esc - should still close the component.

## Screenshots 
template-part-selection-keyboard

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/template-part/edit/selection/index.js [hunk 1]
```diff
@@ -3,15 +3,32 @@
...
+
+// Disable reason (no-static-element-interactions): Navigational key-presses within
+// the menu are prevented from triggering WritingFlow and ObserveTyping interactions.
+/* eslint-disable jsx-a11y/no-static-element-interactions */
 export default function TemplatePartSelection( { setAttributes, onClose } ) {
 	const [ filterValue, setFilterValue ] = useState( '' );
 	return (
-		<>
+		<div
+			onKeyPress={ stopKeyPropagation }
```
Comment (noahtallen) [inline packages/block-library/src/template-part/edit/selection/index.js:29]: I'm curious why we need two different handlers here? Also curious, can we just attach these to `SearchForm`?

**Review by noahtallen — APPROVED** [review]
This works perfectly for me, nice! Just left a couple of notes (I think the package-lock might need to be added before merge). 🚀

File: packages/block-library/package.json [hunk 2]
```diff
@@ -59,6 +59,7 @@
 		"memize": "^1.1.0",
 		"moment": "^2.22.1",
 		"react-easy-crop": "^3.0.0",
+		"reakit": "1.1.0",
```
[the remark this answers is not in the document]
Comment (Addison-Stavlo) [reply packages/block-library/package.json:62]: Now that you mention it, I would think so!  Then looking, there don't seem to be package-locks inside the package directories themselves, just at the top level of the project?

[the remark this answers is not in the document]
Comment (noahtallen) [reply packages/block-library/package.json:62]: yeah, I think it get picked up by `npm install` from the root

[the remark this answers is not in the document]
Comment (noahtallen) [reply packages/block-library/package.json:62]: (at least, that has been the case for me!)

File: packages/block-library/src/template-part/edit/selection/index.js (hunk 1 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/selection/index.js:29]: > I'm curious why we need two different handlers here?

Im not entirely certain myself, but I saw that these were both used in the quick-inserter:
https://github.com/WordPress/gutenberg/blob/c2ea1b27df08117dbf53200630e8a627f4479df1/packages/block-editor/src/components/inserter/quick-inserter.js#L219-L220
My assumption is there are other controls that may fire off keyPress as well that we don't want to trigger from here.

> Also curious, can we just attach these to SearchForm?

Ah yeah, I added it to the wrapper for both before using composite to add controls for the list.  However it doesn't seem to stop the events when put on SearchForm, so Im guessing that component doesn't properly merge/handle these event handlers that are passed down to it.

File: packages/block-library/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Addison-Stavlo) [reply packages/block-library/package.json:62]: Yeah, im confusing myself.  I did `npm ci` and it didn't pick it up.  Trying again as the linter check isn't happy with it either.

[the remark this answers is not in the document]
Comment (Addison-Stavlo) [reply packages/block-library/package.json:62]: There we go.  regular old `npm install` did it 😆

[the remark this answers is not in the document]
Comment (noahtallen) [reply packages/block-library/package.json:62]: Yeah, I think the idea of `npm ci` is to just install straight from the existing package lock without updating anything 😅

File: packages/block-library/src/template-part/edit/selection/index.js (hunk 1 above)
Comment (noahtallen) [reply packages/block-library/src/template-part/edit/selection/index.js:29]: >  My assumption is there are other controls that may fire off keyPress as well that we don't want to trigger from here.

I guess I was wondering more why the two callbacks do slightly different things :)

File: packages/block-library/package.json (hunk 2 above)
Comment (jeyip) [inline packages/block-library/package.json:62]: **Non-blocking:**

Do you happen to know if we have a convention about the caret operator for npm semantic versioning?

**Comment by jeyip** [comment]
**Non-blocking:**
I see this issue in master as well, but there seems to be a problem with quick tab key presses:

1. Open the template selector
2. Upon seeing the dropdown, quickly and immediately press the tab key
3. The dropdown will close

I can reproduce this in Firefox, Safari, Chrome, and Edge.

Non blocking because this seems like an edge case and isn't isolated to this branch.

**Comment by jeyip** [comment]
In Safari, pressing the tab key from the template selector focuses the url search bar instead of the item list.

All other behavior works as described in Chrome, Firefox, Edge, Safari, and IE11

**Comment by jeyip** [comment]
+1 from me whenever the discussion about the Safari bug is resolved 🚀

File: packages/block-library/package.json (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/package.json:62]: Im not sure about any specific convention, but I noticed this package is also missing the caret in the `block-editor`'s package.json so it seems ok to go with the exact versioning for it.

**Comment by Addison-Stavlo** [comment]
> I see this issue in the master as well, but there seems to be a problem with quick tab key presses:

Im not able to reproduce this well.  I can only get it to happen if I hit the 'tab' key multiple times quickly while it is opening.

> In Safari, for me, pressing the tab key from the template selector focuses the url search bar instead of the item list. 

It seems like tabbing is a bit funky on Safari, im having issues with tab nav in many places.  The quick inserter seems to do the same thing.  In our selection component I can hit 'shft+tab' from the input and it will move into the list items.  However, with the quick inserter I cannot access the blocks list that way at all.  🤔

File: packages/block-library/package.json (hunk 2 above)
Comment (jeyip) [reply packages/block-library/package.json:62]: Gotcha. Resolving this conversation because formalizing a convention will be outside the scope of this PR.

**Comment by Addison-Stavlo** [comment]
> It seems like tabbing is a bit funky on Safari,

🤦‍♀️  - Its totally fine in Safari after enabling tab navigation.  Safari is 'funky' in that you need to actually enable it in preferences->advanced:
Screen Shot 2020-10-07 at 12 36 48 PM

**Comment by jeyip** [comment]
> Its totally fine in Safari after enabling tab navigation. Safari is 'funky' in that you need to actually enable it in preferences->advanced:

TIL 🙃 

Thanks for figuring that out @Addison-Stavlo!

**Review by jeyip — APPROVED** [review]
Nice job on this! 💯

**Comment by jeyip** [comment]
> Im not able to reproduce this well. I can only get it to happen if I hit the 'tab' key multiple times quickly while it is opening.

My mistake! I mischaracterized the problem. If you quickly press the tab key _twice_ while the dropdown previews are still loading, the dropdown will close.

**Pull request merged**